trade-finance-manager-ui/server/controllers/utilisation-reports/helpers/reconciliation-for-report-helper.ts

```diff
@@ -222,12 +227,62 @@ const mapPaymentToPaymentDetailsPaymentViewModel = (
 const getPaymentsSortedByDateReceived = (payments: Payment[]): Payment[] =>
   orderBy(payments, [({ dateReceived }) => parseISO(dateReceived).getTime()], ['asc']);
 
+/**
+ * Gets the payment id to date reconciled data sort value map
+ * @param feeRecordPaymentGroups - The fee record payment groups
+ * @returns The payment id to date reconciled data sort value map
+ */
+const getPaymentIdToDateReconciledDataSortValueMap = (feeRecordPaymentGroups: FeeRecordPaymentGroup[]): Record<number, number> => {
+  const allPaymentsWithDateReconciled = feeRecordPaymentGroups.reduce(
+    (paymentsWithDateReconciled, { paymentsReceived, dateReconciled }) => {
+      if (!paymentsReceived) {
+        return paymentsWithDateReconciled;
+      }
+      return [...paymentsWithDateReconciled, ...paymentsReceived.map((payment) => ({ ...payment, dateReconciled }))];
+    },
+    [] as (Payment & { dateReconciled?: IsoDateTimeStamp })[],
+  );
+  return orderBy(allPaymentsWithDateReconciled, [({ dateReconciled }) => (dateReconciled === undefined ? -1 : parseISO(dateReconciled).getTime())]).reduce(
+    (map, { id }, paymentIndex) => ({
+      ...map,
+      [id]: paymentIndex,
+    }),
+    {} as Record<number, number>,
+  );
+};
```

Comment:
Are we able to split this up and/or simplify, there's a lot going on and feels expensive?

```
1) reduce
2) reduce logic
3) reduce returns an array with nested mapping
4) return an array ordered by XYZ with date checks/parsing, then reduced with a mapped object
```

Reply:
The logic you've suggested feels just as expensive as the existing logic and, whilst I agree that this function is fairly complicated, a good amount of dev time over the past months has been dedicated to trying to generate these kinds of sorted fields, and the above logic is a good compromise between readability, maintainability and speed.

I will look into moving this logic to another file though so that this file can be made smaller

Reply:
The logic I pasted wasn't a suggestion it's a summary effectively of the existing logic - would be great if we can simplify this - i.e why do we need to reduce 3-4 times

Reply:
We need to reduce multiple times because of a few reasons:
- the groups are an array
- the `paymentDetails` stored in the group is also an array, hence requiring a reduce to get all the payments
- to convert the array to a map, we need a reduce